docs/how-to/test-dataset-definition.md

```diff
@@ -86,9 +85,20 @@ test_data = {
 }
 ```
 
+In the example above we have created one test patient with the patient ID `1` and added test data for two tables: `patient` and `medications`.
+The keys of the second-level dictionary match the names of the tables in the dataset definition.
+To explore how these tables are structured you can look at the column names in the [table schemas](../../reference/schemas) documentation.
+As mentioned above, adding data is different for *one-row-* and *many-rows-per-patient* tables:
+
+* `patients` is a *one-row-per-patient* table, so you can only define one dictionary with one key for each column (`date_of_birth` and `sex`) that you want to populate.
+    Note that you don't have to specify a value for each column in the underlying table.
+    For example we did not specify `date_of_death` in the dictionary so the column will be missing with the value `None`.
```

Comment:
Out of interest, what happens when the column has a "Never `NULL`" constraint? (e.g. `open_prompt.ctv3_code`).

Reply:
It *just works* without complaining, e.g., [patients.sex](https://docs.opensafely.org/ehrql/includes/generated_docs/schemas/tpp/#patients.sex) also has the 'Never `NULL`' constraint.

Reply:
I'm not sure whether working without complaining is a good idea. We'd like a user's conceptual model of test data to be the same as that of production data.

Reply:
Iain's right.  But fixing that is outside the scope of this ticket.  (New ticket Milan?)

Reply:
yes, good next ticket, I also agree! But it also means that a user always needs to add values for all 'Never NULL' columns (even if they are not used by the user).

Further to the user's conceptual model, I think this improvement is also useful because a user also needs to meet all the other constraints that a column might have, and therefore the test data will match the real data :)

Reply:
I'll write the ticket up

Reply:
Issue with description of the problem here: https://github.com/opensafely-core/ehrql/issues/1858